List activities for subscriptions/digest

Gauravano opened the pull request:
Fixes #2027

jywarren commented:
This looks fantastic, and is so exciting to see! This'll be a really cool page. 

Can we add a simple functional test to it, though? That way we know the page renders in tests. 

Thank you!!!

Gauravano commented:
Hey, @jywarren I have a doubt regarding this, kindly see https://github.com/publiclab/plots2/issues/2027#issuecomment-359258980 .Thanks.

jywarren commented:
I agree. We want all nodes which the user is subscribed to AND which have changed in that time period. 

This doesn't quite work for that:

https://github.com/publiclab/plots2/blob/93358c4b3aa38513e00be226800aa0194a6dce7d/app/models/user.rb#L314

I think we need to wrap this function:

https://github.com/publiclab/plots2/blob/93358c4b3aa38513e00be226800aa0194a6dce7d/app/models/user.rb#L169-L174

How's that sound? Good catch on that method, and if we change it it may also affect the unit test which we could update to match. Thanks so much!

Gauravano commented:
Hi @jywarren, the current method is working but I found something strange. node.changed time is different from node.latest.created_at i.e., from its latest revision.Could you suggest why is it so ? Also, this problem is highlighted from the view as view is using node.latest.created_at time and I have used node.changed time:-
![activity](https://user-images.githubusercontent.com/20878070/35351432-e0f1933e-0166-11e8-9786-7af817bbd4f1.png)

jywarren commented:
Ah, i see. Yes, the two can drift, which is a problem. I think we should be sure to update `.changed` on these lines:

https://github.com/publiclab/plots2/blob/master/app/models/node.rb#L533-L539

Be sure to change `changed` using this syntax (below) since `changed` is a reserved word in Rails, and this field dates from pre-Rails code: 

https://github.com/publiclab/plots2/blob/ceb7d05404b4619c7fa8cd8cae5e8215798cb02c/app/models/node.rb#L111

However, you can also change your query to use `revision.update_at` by adding these lines to the query:

```ruby
                 .includes(:revision, :tag)
                 .references(:node_revision)
```

As on these lines:

https://github.com/publiclab/plots2/blob/ceb7d05404b4619c7fa8cd8cae5e8215798cb02c/app/controllers/wiki_controller.rb#L370-L374

How does that sound? 

Thank you!!!

Gauravano commented:
Actually, @jywarren,node.changed is showing latest time but time shown by revisions is not updated, that's issue.I think its related to #2092.Anyways I'll  try running some more tests and thanks for helping Jeff.

jywarren commented:
Hmm, revisions should be auto set because rails auto sets some timestamp fields. But this is a legacy table so maybe not. The revisions should be getting timestamps in node.new_revision in node.rb. Hope that helps!

Gauravano commented:
Hi @jywarren, I have displayed node.changed time on digest page as we are comparing with it only in the controller , removing all conflicts.Please review it.Thanks.

jywarren commented:
OK, this is looking really fantastic. Can we add a functional test for the `digest` action, though, just to be sure that doesn't break? 

Are you limiting ability to view to only people logged in, with `require_user`? 

Awesome!

Gauravano commented on app/controllers/subscription_controller.rb:
```diff
@@ -5,7 +5,7 @@
 
   respond_to :html, :xml, :json
-  before_filter :require_user, :only => [:create, :delete, :index]
+  before_filter :require_user, :only => [:create, :delete, :index, :digest]
```
Yes @jywarren here is the line for checking current_user

Gauravano commented:
> Are you limiting ability to view to only people logged in, with require_user?
Yes @jywarren its already handled

Gauravano commented:
Hey, @jywarren I have added the functional tests.here's preview : 

![weekly_digest](https://user-images.githubusercontent.com/20878070/35669527-5f50969a-075b-11e8-8c53-b721fb5691b3.gif)

Don't you think we need to change the page heading as we are showing users content that was changed or created with the tag they are subscribed to?

jywarren commented:
This looks great. Yes, sorry i guess we should explain more what the content is -- agreed. Maybe a new PR for that? 

Awesome!

jywarren commented:
"Content from the past week on topics you've subscribed to"?

Gauravano commented:
Sounds good ...
